CRM/Civicase/Service/CaseTypeCategoryBaseEventHandler.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+use CRM_Civicase_Service_CaseTypeCategoryProcessor as CaseTypeCategoryProcessor;
+
+/**
+ * Class CRM_Civicase_Service_CaseTypeCategoryBaseEventHandler.
+ */
+class CRM_Civicase_Service_CaseTypeCategoryBaseEventHandler {
+
+  /**
+   * Run some OptionValue actions on some event.
+   *
+   * @param string $op
+   *   The operation being performed.
+   * @param string $objectName
+   *   Object name.
+   * @param mixed $objectId
+   *   Object ID.
+   * @param object|array $objectRef
+   *   Object reference.
+   * @param bool $loadFullData
+   *   (Optional) If TRUE then full case type category data will be loaded
+   *   from db. Is useful if $objectRef contains some partial data.
+   */
+  public function run($op, $objectName, $objectId, &$objectRef, $loadFullData = FALSE) {
+    if (!$this->shouldRun($objectName, $objectRef)) {
+      return;
+    }
+
+    $processor = new CaseTypeCategoryProcessor($this->normalizeCategory($objectId, $objectRef), $loadFullData);
+    if ($op == 'create') {
+      $processor->createMenuItem();
+    }
+    elseif ($op == 'edit') {
+      $processor->updateMenuItem();
+    }
+    elseif ($op == 'delete') {
+      $processor->deleteMenuItem();
+    }
```

Comment:
Lumping the create/edit and delete functions in one run function has defeated the initial purpose of the abstraction we had in mind.